### ISSUE TITLE
Upgrade to using Babel over Traceur for compiler transformers

### ISSUE THREAD

**Issue by guybedford** [systemjs/builder#489]
All the transformers in https://github.com/systemjs/builder/tree/master/compilers should be upgraded to be based on Babel transformation over Traceur transformations.

This would then support comments in the output source.

**Comment by guybedford** [comment]
In addition this will better support source maps.

**Comment by dschnare** [comment]
Is there anyway to have the `transpiler` config option take effect instead of having the transpiler hardcoded to Babel/Traceur? Using `System.config()` then `System.import()` it respects the custom `transpiler`, but `builder.config()` seems to have difficulty tracing the dependency graph for ES6 modules when I have a custom transpiler specified.

If it helps I was attempting to use the Typescript compiler.

**Comment by guybedford** [comment]
It's much better to use http://github.com/frankwallis/plugin-typescript over the internal TypeScript transpiler.

**Comment by dschnare** [comment]
Thanks, I'll give that a shot.

**Comment by rjharmon** [comment]
I'm left confused by the various tickets related to this.  While I don't see that it's solved, I don't see any tickets still open for solving it.  a) is it actually solved?  b) if so, what package versions would have the improvements, and c) if not, what open issue(s) remain (or does one or more of the existing closed issues need to be re-opened?)?

Thanks!

**Comment by guybedford** [comment]
This is entirely to do with SystemJS builder's internal transformation code and not to do with which ES6 transpiler is being used.

**Comment by jrauschenbusch** [comment]
Hi @guybedford,

can you give a rough estimate when the upgrade to Babel for the Internal transformation code will be completed?

Thanks for your great work on jspm, systemjs and systemjs builder!

**Comment by guybedford** [comment]
@jrauschenbusch it's not currently a priority unfortunately.

**Comment by asapach** [comment]
@guybedford when will this become a priority? The source maps issue is a show-stopper for production use.
Is this something the community can help with?

**Comment by guybedford** [comment]
@asapach the task here is simply to rewrite the files in https://github.com/systemjs/builder/tree/master/compilers as Babel transforms. The tests cover probably about 70% of the cases, with the rest of the details in the code itself, so I would gladly provide feedback if someone wants to take this on before I can get to it.

**Comment by asapach** [comment]
OK, I'll give it a try. Which one is the simplest one to start with in your opinion?

**Comment by guybedford** [comment]
https://github.com/systemjs/builder/blob/master/compilers/register.js is the simplest.

**Comment by asapach** [comment]
I've created the transformer in a separate repo: https://github.com/asapach/babel-plugin-transform-system-register. I'm not sure where it will end up, but it was easier for me to write it test-first that way. (BTW if there are any test cases I've missed please let me know.)
I also have it inline in a branch: https://github.com/asapach/builder/blob/transformer/compilers/register.js

Good news:
* Source maps are *beautiful*
* The code is simpler (and covered by tests)

Not-so-good news:
* Babel has different heuristics for dealing with source maps:
  * It doesn't derive `sourceRoot` automatically
  * If there are input source maps it seems to ignore all the options and modifies the existing source map rather than creating a new one
  * In the same fashion it does not include the intermediary file into the resulting source map (e.g. `foo ← bar` will only contain `foo`)
  * It doesn't normalize the paths into URLs
  * Basically garbage in - garbage out.

Based on this some of the source map related logic might need some adjustment.

Feedback is welcome.

**Comment by guybedford** [comment]
@asapach this looks amazing!! Thank you so much for working on this :)

I'm happy to keep it as a separate project if you like, as I agree it's a nice encapsulation.

We do use babel input source maps against file URLs in https://github.com/systemjs/plugin-babel/blob/master/plugin-babel.js#L169 against file URLs and it seemed to work ok for me, unless these are additional edge cases that would apply there too?

I guess we just need to be clear that the assumption expected by the builder is to get back a source map that uses file: URLs and no sourceRoot. Or we can ensure we do a more consistent sourceRoot sanitization.

Perhaps just a post-processing stage to turn the sources back into file URLs would be enough. It shouldn't be dropping relations though, that sounds odd to me. Was `bar` being dropped because it was referenced via a sourceMappingURL comment or something? Or was it a truly flattened source map to begin with?

**Comment by asapach** [comment]
I ran it on the test suite: for example there's [`register.js`](https://github.com/systemjs/builder/blob/master/test/fixtures/test-tree/register.js) that has a source map that points to [`babel.js`](https://github.com/systemjs/builder/blob/master/test/fixtures/test-tree/babel.js). If you bundle `register.js` with traceur you will get both `register.js` and `babel.js` in `sources`. The former one will have very few mappings, but it will still end up there. If you run it with babel, `register.js` will be missing entirely, because it doesn't actually contribute anything to the mapping. I ran a similar test on my production project and I get similar results.

**Comment by guybedford** [comment]
There's no reason to have 'register.js' in the new source map as it is only
an intermediate file, so it sounds current to me, and the Traceur
implementation seems wrong?

**Comment by asapach** [comment]
Agree. But that's a potentially breaking change.

Should I create a PR for this?

**Comment by guybedford** [comment]
A PR would be amazing, thank you! I think a change in source maps behavior
doesn't have to be a major version change?

**Comment by asapach** [comment]
OK, I'll publish the babel transformer to npm and submit the PR on the weekend.
It doesn't really change the api, so technically it's not a major change, but expect some bug reports :)

What is the next transformer you would like me to look at?

**Comment by asapach** [comment]
Global transformer looks easy, I'll take it next.

**Comment by jrauschenbusch** [comment]
Great to see this activity here. Am I correct that the json compiler does not require any change?

**Comment by asapach** [comment]
JSON compiler doesn't use traceur (or support source maps for that matter), so it doesn't have to change.

**Comment by jrauschenbusch** [comment]
Ok. So the next transformer should be cjs.js due to the dependence of amd.js to cjs.js, right? I think that the asm.js transformer is the most complex one.

**Comment by jrauschenbusch** [comment]
Like @asapach i've implemented the require remap transformer part of cjs.js in a separate repo:
https://github.com/jrauschenbusch/babel-plugin-transform-remap-require

I think this small plugin could be used for any function call remapping tasks, but for now it should be dedicated to the transformer upgrade.

I'll go on with the rest of the cjs transformer implementation. I hope this is useful for the upgrade process to move forward.

**Comment by guybedford** [comment]
@jrauschenbusch nice - that looks like it may well be useful.

**Comment by asapach** [comment]
Created the PR: #652

**Comment by jrauschenbusch** [comment]
Great to see the first PR. My further work related to the cjs transformer is now online, but it's still WIP!
https://github.com/jrauschenbusch/babel-plugin-transform-cjs-to-systemjs

Things to do:
1.) Understand what hasRemoveUseStrict(list) does and re-implement it for the babel plugin
2.) Fixing the first parameter for `SystemJS.registerDynamic`:
  * Original: https://github.com/systemjs/builder/blob/master/compilers/cjs.js#L178
  * Babel: https://github.com/jrauschenbusch/babel-plugin-transform-cjs-to-systemjs/blob/master/src/index.js#L107
  * Result: `SystemJS.registerDynamic('{}', ...);`

3.) `inherit` the require-remap logic from https://github.com/jrauschenbusch/babel-plugin-transform-remap-require
4.) Implemeting adaquate tests

**Comment by asapach** [comment]
Here's the second PR: #653. It's on top of the other one, so might be a bit noisy.

**Comment by jrauschenbusch** [comment]
@asapach Nice job! Looks very clean and well tested! With which transformer you will continue?

**Comment by asapach** [comment]
I think I'll pause for a bit and wait for @guybedford to merge the changes. Once a new version in released I'll do some thorough testing before continuing.
Both CJS and AMD transformers seem pretty complicated and interconnected.
I see that ESM compiler is already using babel - is there more work to be done?

**Comment by jrauschenbusch** [comment]
Hey @asapach, this is a good plan to go on.

Indeed, the AMD transformer is using the exported CJSRequireTransformer part from the CJS transformer. Hence i've outsourced this part as a separate babel plugin to be able to reuse it. The rewrite of the CJS transformer is not so complicated as it first seems (nearly same complexity as your global wrapper plugin). But i've to say that your plugins are much cleaner as mine. Maybe you can give a look into it and provide a cleaner way as PR.

Regarding the ESM transformer i think we only need a cleanup from legacy code. @guybedford: can you provide a more precise answer to this question?

**Comment by guybedford** [comment]
It looks like we're nearly ready to merge the global transformer. Apart from that it is just AMD and CJS - yes the ESM one is already running through Babel, with some backwards-compatibility cases for old SystemJS internal compilation which can be deprecated with the next SystemJS major.

**Comment by asapach** [comment]
I've noticed one more thing as I was digging through the other module plugins: I think we could use the existing Babel options for module names and normalization:
```js
{
  moduleIds: !opts.anonymous,
  moduleId: !opts.anonymous && load.name,
  resolveModuleSource: function(dep) {
    if (opts.normalize)
      return load.depMap[dep];
    else
      return dep;
  }
}
```
Instead of the custom plugin options `moduleName` and `map`.

**Comment by jrauschenbusch** [comment]
I've cleaned up the CJS transformer and given a new name to it:
https://github.com/jrauschenbusch/babel-plugin-transform-cjs-system-wrapper

It has an internal dependency to https://github.com/jrauschenbusch/babel-plugin-transform-cjs-system-require. Also i've created a release of the transform-cjs-system-require plugin in GitHub, but it's not published yet in npm. Hence to be able to test it, one need to manually do a `npm pack` (tgz output) and install it locally in the cjs-system-require path by `npm i /path/to/babel-plugin-transform-cjs-system-require-0.0.1.tgz`.

Feedback is welcome!

Regarding the module name & normalization: Can you provide an example how you would implement it?

**Comment by asapach** [comment]
> Can you provide an example how you would implement it?

Other plugins seem to be doing it this way: https://github.com/babel/babel/blob/a8fbc98a6df6a7d3d141eab05567a6fcba0d7311/packages/babel-plugin-transform-es2015-modules-systemjs/src/index.js#L231

**Comment by jrauschenbusch** [comment]
Ok, this could be an alternative to make it still cleaner.

Currently i'm not so happy with the line feeds in the CJS transformer inserted by babel. Maybe one has a solution for this? The idea was to replace each WILDCARD if it has a `value !== null` and else to remove the WILDCARD not producing an empty line. But this doesn't seem to work.

**Comment by jrauschenbusch** [comment]
@asapach As it seems `resolveModuleSource` is only called for import statements not for require statements.

**Comment by asapach** [comment]
yeah, I think I meant `getModuleId()`:
https://github.com/babel/babel/blob/4c371132ae7321f6d08567eab54a59049e07f246/packages/babel-core/src/transformation/plugin-pass.js#L33
https://github.com/babel/babel/blob/dc1f40540de11045bb4714350350e012fee4f93e/packages/babel-core/src/transformation/file/index.js#L202

**Comment by guybedford** [comment]
@asapach I'm not quite sure what you mean. That is the approach we use in the ES module compiler already I believe.

**Comment by asapach** [comment]
I mean that instead of passing [custom plugin options](https://github.com/systemjs/babel-plugin-transform-system-register#via-node-api-recommended) `moduleName` and `map` we could try using [standard babel options](https://babeljs.io/docs/usage/options/) `moduleId` and `getModuleId` correspondingly.

**Comment by guybedford** [comment]
Ah right, so as a convention? That would make sense.

**Comment by guybedford** [comment]
@asapach for AMD transformation, it's long been on the list to handle this as a full wrapping, see https://github.com/systemjs/builder/issues/264 for more info. If you have any ideas here and want to discuss further just let me know. Otherwise the direct copying may well make sense too!

**Comment by asapach** [comment]
It would be good to have some examples / test cases to understand what you mean by full wrapping. Maybe add some details to #264?

**Comment by guybedford** [comment]
@asapach I've expanded on the idea in https://github.com/systemjs/builder/issues/264. But it is a little complex to explain. How about I start by putting together all the test cases clearly in a PR over the next week, and then get back to you to discuss further?

**Comment by jrauschenbusch** [comment]
@guybedford How we want go on with CJS transformer? Should i also create a PR?

**Comment by guybedford** [comment]
The CJS transformer is well-defined yes, so whatever you and @asapach agree is best here.

**Comment by asapach** [comment]
@jrauschenbusch I'm going to take a look at your code when I have the time.

**Comment by jrauschenbusch** [comment]
@asapach Alright. You can have a look into my upgraded builder fork with it's last commit: https://github.com/jrauschenbusch/builder

Since the `babel-plugin-transform-cjs-system-wrapper` ist not published yet in npm you have to manually install both, the `cjs-system-wrapper` and `cjs-system-require` modules it via `npm pack` if you want to run the tests.

**Comment by asapach** [comment]
As promised I've done some testing and I haven't seen any issues with the new transformers, so I think it's safe to continue with CJS and AMD. However I'm still not happy with source maps - I see some inconsistencies when stepping through code: lines jumping around, unable to set a breakpoint, if/else blocks not being triggered, etc. I need to isolate the problem, so I'll spend some more time on this.

@jrauschenbusch: your code looks good, I've sent a couple of PRs.

**Comment by jrauschenbusch** [comment]
@asapach Thanks for your detailed review! I've merged all of the PRs where no further discussions are needed. Two issues need some discussion to proceed. After solving this issues/PRs i would release the transformer packages officially in npm, make a PR for the builder integration and transfer the ownership to the repositories to the SystemJS ecosystem.

**Comment by jrauschenbusch** [comment]
@asapach jochen.rauschenusch@gmail.com the packages are released and the PR is open.

**Comment by jrauschenbusch** [comment]
So the last one is the AMD transformer which is not an easy task. It is immediately connected with issue #264. Maybe it would help to define some unit tests before we go into implementation details?!

Like @guybedford i also think wrapping AMD modules in System.registerDynamic(....) would be the best way to carry on.

**Comment by guybedford** [comment]
@jrauschenbusch apologies to be the bottleneck on the last one. I'll put some time aside on Monday to do this if that is ok? I'm just working on loader updates this week as the spec changed.

**Comment by jrauschenbusch** [comment]
@guybedford Nothing to apologize here. Your plan is absolutely fine for me! If there's an essential feature that temporarily needs more of your efforts, the rest just must wait.

**Comment by guybedford** [comment]
@jrauschenbusch I started looking into the AMD test cases, but as posted at the issue there it doesn't work well for multi-define files so may be worth sticking with what we have.

Is there enough information to go by to follow the current AMD transformation exactly, or do you need any clarifications on how it works?

**Comment by jrauschenbusch** [comment]
@guybedford I also think we should start with what we currently have. I'll take a detailed look into the current amd transformation process. But first results will be earliest available in two weeks since i'm on a holiday trip starting tomorrow. :sunglasses:

**Comment by guybedford** [comment]
@jrauschenbusch let me take one more look at it this afternoon based on the current suggestions and get back to you further on the approach. Thanks again so much for offering to help here.

**Comment by jrauschenbusch** [comment]
@guybedford I think this issue could be closed as #764 is now merged into 0.16 branch. Or should we still wait unitl the last traceur fragments are removed from the code base?

**Comment by guybedford** [comment]
A huge thanks to @asapach and @jrauschenbusch for completing this work.
